operations/kaas/cluster-stack-operator.md

```diff
@@ -0,0 +1,692 @@
+
+In sum, this controller embodies a fusion of precision, adaptability, and user-centricity. By ensuring meticulous version tracking, adept handling of operational nuances, and continuous user feedback, it exemplifies the pinnacle of Kubernetes management prowess.
+
+
+### ProviderClusterStackRelease controller - Example
+
+Delving into the intricacies of this provider's operations, let's visualize its underlying processes that seamlessly ensure the user gets the specified node images uploaded as per their requirements.
+
+1. **The Initial Step - Creating ProviderNodeImageRelease Objects:**
+The `ProviderClusterStackRelease` plays a foundational role by birthing new `ProviderNodeImageRelease` objects. This creation is not random but carefully aligned with each node image delineated in the spec. It’s like having a dedicated steward for each image the user specifies.
+
+2. **Interlinking Statuses in the Control Loop:**
+The control loop is not just about overseeing operations. It actively fetches the status of each `ProviderNodeImageRelease` object. Remember, each of these objects shoulders the responsibility of ensuring a particular node image becomes accessible to the user. The loop captures this status and etches it into the `ProviderClusterStackRelease` object’s status. It’s akin to having an automated status update system ensuring synchronicity across different processes.
+
+3. **The Determination of Readiness:**
+Once all node images reach a state of preparedness, the controller makes its proclamation by setting `status.ready: true`. This is the green signal for the clusterstackrelease-controller to swing into action and proceed with its tasks.
+
+4. **Safeguarding Node Images with a Dedicated Controller:**
+The `ProviderNodeImageRelease` object doesn't operate in isolation. It has its own guardian angel, a dedicated controller, ensuring that a specific node image is primed for the user. This is no simple task. The path to readiness often involves time-consuming operations, such as the meticulous uploading or intricate building of the node image.
+
+5. **Tackling the Challenge of Long-Running Operations:**
+These lengthy operations come with their own set of hurdles. Consider scenarios where the operator's container is abruptly halted by Kubernetes or manually terminated. The solution? A robust map securely stored in memory, vigilantly monitoring all concurrent operations (thankfully, the power of goroutines ensures parallel operations). Handling shutdowns with finesse, especially those triggered by SIGTERM, becomes paramount. The key lies in gracefully stopping all extended operations, potentially leveraging the canceling contexts in Go.
```

Comment:
As a side note as for another controller I recently faced this exact issue again. There's no reliable way to execute shutdown actions within k8s even for pods under certain scenarios. Therefore the mentioned finesse must be added while handling with the data, not at shutdown time.

Reply:
agreed, the controller cannot do anything when it is killed. We then have to either handle some form of cleanup outside of the operator or when the operator re-starts.

Reply:
Ideally, there should not be any long-running operations in controllers at all, right?

If an operation really is long-running, tracking progress using usual (mostly stateless) reconciliation loops (setting `status` to track preliminary states) should be the K8s native way to do it right? Maybe even splitting out some things into their own resource definitions with their own reconciliation loops?

That may indeed prove hard to do if some long-running synchronous call has to be executed - but if that is not something that is already foreseeable, IMHO this probably is not something to worry about beforehand.

Reply:
Agreed that there shouldn't be long-running operations at all. But here we cannot avoid it, as building or uploading images might be long-running. In this case we use goroutines that work in parallel and that don't block the reconciliation loop. 

The only other alternative is starting separate containers via Kubernetes jobs, but that would be quite messy as well.

Reply:
Creating a Kubernetes job might appear simple initially, but it can lead to more overhead than anticipated. Here's why:

At first glance, creating a job in Kubernetes seems efficient because it manages the lifecycle of your task. But, there are challenges:

1. Monitoring: If the job or the pod fails, how would you know? For reliability, you'd need a monitoring solution to check the job's health.
1. Error Recovery: If a job gets stuck, you would need logic to restart it.
1. Data Cleanup: If your system identifies images by their names, you might end up with unused or "orphaned" images. A garbage collector would be needed to clean them up.
1. State Information: You would also need a way to regularly check the status or progress of the pod. This could involve accessing the pod directly or having the job send status updates.
1. User Visibility: To make the system user-friendly, you'd need a way to show users the progress or status of their tasks.

Considering these complexities, it might be simpler to directly run a pod and manage its lifecycle with a controller. However, this introduces a testing challenge. Unit tests may not be sufficient, and tools like envtest can't handle this scenario. Therefore, you'd rely on end-to-end (e2e) tests. These tests are slower, tougher to maintain, and can be time-consuming when identifying issues.

You might then wonder: "Why run this as a separate pod at all?" Is it to estimate resource needs or to integrate more smoothly with other systems?

An alternative is using Go routines. Here are their benefits:

1. Easy Testing: You can test the logic easily, and native tools work out-of-the-box.
1. Simplified Management: The lifecycle is managed directly in the code. There's no need to interact with the Kubernetes API or set up complex RBAC.
1. Real-Time Information: Getting updates about the ongoing task is straightforward.
1. Resource Management: Go routines can be queued, helping predict resource usage. Since in our use case one controller won't spawn hundreds of goroutines simultaneously, it's efficient and avoids unnecessary overhead.

Conclusion: If you're aiming for a quick fire and forget method, a Kubernetes job is suitable. However, for building robust systems, alternative methods like Go routines are more straightforward and reliable. They offer precise control over long-running tasks, ensure easier maintainability, and reduce the intricacies of testing. This leads to better code quality in less time. Of course, using an approach with creating pods is entirely feasible. If we were operating on a larger scale, this would indeed be the preferred solution. However, it comes with all the previously mentioned challenges, adding unnecessary overhead for our current needs. As the saying goes, usually the simplest solution is the best! ;)

Reply:
> But, there are challenges

As always :D 

> As the saying goes, usually the simplest solution is the best! ;)

I can also agree to this statement, while I still think that it actually favours a `Job` approach. Mirroring your points:

> Monitoring: If the job or the pod fails, how would you know? For reliability, you'd need a monitoring solution to check the job's health.

A monitoring solution is not strictly required, as [`JobStatus v1`](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.25/#jobstatus-v1-batch) includes `conditions` (`Failed`/`Complete`).

> Error Recovery: If a job gets stuck, you would need logic to restart it.

For clean exits, this is actually builtin (and the primary purpose of a Job compared to a one-off Pod): "A Job creates one or more Pods and will continue to retry execution of the Pods until a specified number of them successfully terminate." (https://kubernetes.io/docs/concepts/workloads/controllers/job/ CC BY 4.0).
For cases, where a non-zero-error exit is not done by the process on its own (-> being "stuck"), I guess using `livenessProbes` would be the way to go, forcing a non-zero-error return.

> Data Cleanup: If your system identifies images by their names, you might end up with unused or "orphaned" images. A garbage collector would be needed to clean them up.

Yes, cleaning up is required, indeed (even though some cleaning up will also be required when implementing in-controller-process-jobs, if leakage should be prevented). If not done in the controller itself, the builtin TTL for Jobs can be used.

> State Information: You would also need a way to regularly check the status or progress of the pod. This could involve accessing the pod directly or having the job send status updates.

Querying [`JobStatus v1`](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.25/#jobstatus-v1-batch) should be sufficient, I guess.

> User Visibility: To make the system user-friendly, you'd need a way to show users the progress or status of their tasks.

That indeed would not be something that is builtin into Kubernetes already, but I am also unsure how something like progress reporting (or progress bars) would be implemented when using in-controller-process-jobs. Would the controller update `status` of Cluster Stack Releases, notifying about stages of image building and pushing?

---

Regarding the use of Go routines for implementing in-controller-process-jobs: I like Go routines very much as well and have a fair share of experience using them, but here are a few disadvantages to this approach:

1. One would still have to implement a non-trivial lifecycle system - which is easier to do in Go than in most other languages, indeed. But it will still be a non-trivial problem implemented in a stateful process that may be terminated during e.g. node reboots. So, if any bug appears...
2. ...identifying a problem is much more complicated for anyone who does not have a Go debugger (or dev environment at least) at hand in e. g. production. When having almost all state inside of K8s API objects, problems are found much easier - for anyone familiar with K8s, but not necessarily with Go, Go routines and the to-be-built lifecycle system on top.

So, at least this point:

> Simplified Management: The lifecycle is managed directly in the code. There's no need to interact with the Kubernetes API or set up complex RBAC.

...has a large flip side. I think that interacting with the K8s API is actually the most natural fit for any controller/operator, and that RBAC actually is not too complex most of the time.

Privilege escalation via Job/Pod creation could be a real tradeoff, though.

Reply:
> Privilege escalation via Job/Pod creation could be a real tradeoff, though.

I guess that this strongly relates also to @fdobrovolny's [remark](https://github.com/SovereignCloudStack/docs/pull/50/files#r1301932822) about the image building process being relatively unspecified.

Building images inside of a container is in my experience usually a rather complicated task, when it comes to security and isolation. The controller may need more privileges on the host system or may schedule e. g. dedicated one-off kaniko Pods. That would be still subject of research/decision making, I guess?